numba_dpex/experimental/launcher.py

```diff
@@ -0,0 +1,297 @@
+):
+    dref = sycl.dpctl_queue_get_device(builder, queue_ref)
+    cref = sycl.dpctl_queue_get_context(builder, queue_ref)
+    args = [
+        cref,
+        dref,
+        kernel_bc,
+        llvmir.Constant(llvmir.IntType(64), kernel_bc_size_in_bytes),
+        builder.load(create_null_ptr(builder, targetctx)),
+    ]
+    kbref = sycl.dpctl_kernel_bundle_create_from_spirv(builder, *args)
+    sycl.dpctl_context_delete(builder, cref)
+    sycl.dpctl_device_delete(builder, dref)
+
+    return kbref
+
+
+def _get_num_flattened_kernel_args(
+    kernel_targetctx: DpexKernelTargetContext,
+    kernel_argtys: tuple[types.Type, ...],
+):
+    num_flattened_kernel_args = 0
+    for arg_type in kernel_argtys:
+        if isinstance(arg_type, DpnpNdArray):
+            datamodel = kernel_targetctx.data_model_manager.lookup(arg_type)
+            num_flattened_kernel_args += datamodel.flattened_field_count
+        elif arg_type == types.complex64 or arg_type == types.complex128:
+            num_flattened_kernel_args += 2
+        else:
+            num_flattened_kernel_args += 1
+
+    return num_flattened_kernel_args
+
+
+def _create_kernel_launcher_body(
+    codegen_targetctx: cpu.CPUContext,
+    kernel_targetctx: DpexKernelTargetContext,
+    builder: llvmir.IRBuilder,
+    indexer_argty: Union[RangeType, NdRangeType],
+    kernel_argtys: tuple[types.Type, ...],
+    kernel_module: _KernelModule,
+    index_space_arg: llvmir.BaseStructType,
+    kernel_args: [llvmir.Instruction, ...],
+):
+    klbuilder = kl.KernelLaunchIRBuilder(kernel_targetctx, builder)
+
+    if config.DEBUG_KERNEL_LAUNCHER:
+        cgutils.printf(
+            builder, "DPEX-DEBUG: Inside the kernel launcher function\n"
+        )
+
+    kernel_bc_byte_str: llvmir.Constant = codegen_targetctx.insert_const_bytes(
+        builder.module,
+        bytes=kernel_module.kernel_bitcode,
+    )
+
+    num_flattened_kernel_args = _get_num_flattened_kernel_args(
+        kernel_targetctx=kernel_targetctx, kernel_argtys=kernel_argtys
+    )
+
+    # Create LLVM values for the kernel args list and kernel arg types list
+    args_list = klbuilder.allocate_kernel_arg_array(num_flattened_kernel_args)
+    args_ty_list = klbuilder.allocate_kernel_arg_ty_array(
+        num_flattened_kernel_args
+    )
+    kernel_args_ptrs = []
+    for arg in kernel_args:
+        ptr = builder.alloca(arg.type)
+        builder.store(arg, ptr)
+        kernel_args_ptrs.append(ptr)
+
+    # Populate the args_list and the args_ty_list LLVM arrays
+    klbuilder.populate_kernel_args_and_args_ty_arrays(
+        callargs_ptrs=kernel_args_ptrs,
+        kernel_argtys=kernel_argtys,
+        args_list=args_list,
+        args_ty_list=args_ty_list,
+        datamodel_mgr=kernel_targetctx.data_model_manager,
+    )
+
+    if config.DEBUG_KERNEL_LAUNCHER:
+        cgutils.printf(
+            builder, "DPEX-DEBUG: Populated kernel args and arg type arrays.\n"
+        )
+
+    qref = _get_queue_ref_val(
```

Comment:
What happens if we don't pass any dpnp arrays and allocates them inside dpjit?

Reply:
That is a good point. Somewhere we need to check if the array args are `DpnpNdArray`. Existing `JitKernel` has a check that I think I have not added anywhere.

Reply:
I have added a check to evaluate queue equivalence before we compile.